examples/aflow/benchmark/hotpotqa.py

```diff
@@ -0,0 +1,212 @@
+import string
+import re
+import os
+from datetime import datetime
+
+def is_number(text: str) -> bool:
+    try:
+        float(text)
+        return True
+    except ValueError:
+        return False
+
+def normalize_answer(s):
+    """
+    Normalize answers for evaluation.
+    """
+
+    def remove_articles(text):
+        return re.sub(r"\b(a|an|the)\b", " ", text)
+
+    def white_space_fix(text):
+        return " ".join(text.split())
+
+    def remove_punc(text):
+        exclude = set(string.punctuation)
+        return "".join(ch for ch in text if ch not in exclude)
+
+    def lower(text):
+        return text.lower()
+
+    return white_space_fix(remove_articles(remove_punc(lower(s))))
+
+def calculate_score(ground_truth: str, prediction: str):
+    """
+    Compute the F1 score between prediction and ground truth answers.
+    """
+    prediction_tokens = normalize_answer(prediction).split()
+    ground_truth_tokens = normalize_answer(ground_truth).split()
+    common = Counter(prediction_tokens) & Counter(ground_truth_tokens)
+    num_same = sum(common.values())
+    if num_same == 0:
+        return 0, prediction
+    precision = 1.0 * num_same / len(prediction_tokens)
+    recall = 1.0 * num_same / len(ground_truth_tokens)
+    f1 = (2 * precision * recall) / (precision + recall)
+    return f1, prediction
+
+def ensure_log_file_exists(path: str):
+    log_file = os.path.join(path, 'log.json')
+    if not os.path.exists(log_file):
+        with open(log_file, 'w', encoding='utf-8') as f:
+            json.dump([], f, indent=4, ensure_ascii=False)
+
+def log_mismatch(problem: str, expected_output: float, prediction: str, predicted_number, path):
+    log_data = {
+        "question": problem,
+        "right_answer": expected_output,
+        "model_output": prediction,
+        "extracted_output": predicted_number
+    }
+
+    log_file = os.path.join(path, 'log.json')
+
+    # 检查log文件是否已经存在
+    if os.path.exists(log_file):
+        # 如果存在，加载现有的日志数据
+        with open(log_file, 'r', encoding='utf-8') as f:
+            try:
+                data = json.load(f)
+            except json.JSONDecodeError:
+                data = []
+    else:
+        # 如果不存在，创建一个新的日志列表
+        data = []
+
+    # 添加新的日志记录
+    data.append(log_data)
+
+    # 将数据写回到log.json文件
+    with open(log_file, 'w', encoding='utf-8') as f:
+        json.dump(data, f, indent=4, ensure_ascii=False)
+
+async def load_data(file_path: str, specific_indices: List[int] = None) -> List[dict]:
+    data = []
+    # 异步读取文件内容
+    async with aiofiles.open(file_path, mode="r", encoding='utf-8') as file:
+        async for line in file:
+            data.append(json.loads(line))
+
+    # 然后在随机选择的样本中基于特定索引列表进行进一步筛选
+    if specific_indices is not None:
+        filtered_data = [data[i] for i in specific_indices if i < len(data)]
+        return filtered_data
+
+    return data
+
+def save_results_to_csv(results: List[Tuple[str, str, str, str, float, float]], path):
```

Comment:
I've noticed the save_results_to_csv function is duplicated across several scripts. To avoid redundancy, I recommend moving this function into a shared utils.py module. This will ensure easier maintenance and consistent updates.